### ISSUE TITLE
Tracking Issue for get_many_mut

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#104642]
Feature gate: `#![feature(get_many_mut)]`

This is a tracking issue for `get_many_mut` and `get_many_unchecked_mut`, which provide `&mut T` access to multiple distinct slice elements.

### Public API

```rust
impl [T] {
    pub unsafe fn get_many_unchecked_mut<const N: usize>(&mut self, indices: [usize; N]) -> [&mut T; N];
    pub fn get_many_mut<const N: usize>(&mut self, indices: [usize; N]) -> Result<[&mut T; N], GetManyMutError<N>>;
}

pub struct GetManyMutError<const N: usize> { /* private */ }
```

### Steps / History

- [x] Implementation:
    - Initial implementation in #83608, also received extensive discussion
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- What should the error type / return type look like?
  -  See https://github.com/rust-lang/rust/pull/83608#issuecomment-914619139, https://github.com/rust-lang/rust/pull/83608#issuecomment-861382045, and other comments on that PR
- Should the interface support non-usize indices (e.g., `get_many_mut([0..3, 4..6])`)?
- What are the requirements for the interface? Do we sort the passed array, and then check distinctness, or check in `O(n^2)` time?

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by shepmaster** [comment]
Is there a reason that `core::slice::GetManyMutError` isn't re-exported as `std::slice::GetManyMutError`?

**Comment by orlp** [comment]
Since `N` is a compile-time constant anyway I would suggest adding a check in `get_many_check_valid` that switches to `sort_unstable` for, say, N > 10. This code gets optimized out entirely for `N <= 10`, and vice versa.

```rust
fn get_many_check_valid<const N: usize>(indices: &[usize; N], len: usize) -> bool {
    let mut valid = true;
    if N <= 10 {
        // NB: The optimzer should inline the loops into a sequence
        // of instructions without additional branching.
        for (i, &idx) in indices.iter().enumerate() {
            valid &= idx < len;
            for &idx2 in &indices[..i] {
                valid &= idx != idx2;
            }
        }
    } else {
        let mut sorted = *indices;
        sorted.sort_unstable();
        for i in 1..N {
            valid &= sorted[i-1] < sorted[i];
        }
        valid &= sorted[N-1] < len;
    }

    valid
}
```

On another note, perhaps we should have a `DisjointIndices<const N: usize>` type where `DisjointIndices::new` calls `get_many_check_valid`. Then indexing with `DisjointIndices` does not need to repeat the check, allowing you to amortize the cost of checking disjointness over many operations without needing `unsafe` code.

**Comment by ajwock** [comment]
I also like the idea of DisjointedIndices for this.  And since a pair of indices is likely a common use case for this, perhaps something that could be available would be something like disjoint_pair(lhs: usize, rhs: usize) -> DisjointIndices<2> (panics if lhs == rhs).

**Comment by pervognsen** [comment]
I've tried to use this a few times now. The fact that the current placeholder error type doesn't distinguish out-of-bounds vs duplicate indices makes it hard to even test in something resembling a real use case. For example, if you wanted to implement slice::swap via slice::get_many_mut, you'd want the duplicate index case to be a nop, but you'd still want the out-of-bounds case to panic.

Related: Would it make sense to have an implementation of IndexMut for [usize; N] that panics in the duplicates case? The duplicate indices issue seems like it falls into the same bucket of dynamic borrow conflicts you have with RefCell and RefCell::borrow/borrow_mut (which panics unlike try_borrow/try_borrow_mut), so it seems like a reasonable behavior when you have an upstream invariant that is expected to guarantee uniqueness but you still need the dynamic check for safety.

**Comment by kupiakos** [comment]
For embedded use cases, I would strongly prefer we _not_ add more panics, especially to a fallible `get` method.

**Comment by kupiakos** [comment]
For embedded use cases, I'd strongly prefer we limit the number of panics added by this feature, limiting it only to .

> Would it make sense to have an implementation of IndexMut for [usize; N] that panics in the duplicates case? 

There's not really a way to implement `IndexMut` for multiple indices - what would be the `Output` type? If the `Output` type is `[&mut T]` then you would need some place to store it for `index_mut` to return `&mut Self::Output`.

The same applies to `.get_mut` - it would be awesome to write `let [e0, e10] = foo.get_mut(DisjointIndices::new([1,2]).unwrap())` but there's no way to do it.

**Comment by mcmah309** [comment]
There is also this crate that introduces some macros and methods for retrieving multiple mutable indices from a mutable slice https://crates.io/crates/indices . It uses insertion sort so more optimized towards a small number of indices and already sorted indices. It also has macros that assume the input is already sorted, requiring no sorting at all.

**Comment by tgross35** [comment]
If there isn't any specific reason not to do this, I think we should change to an interface based on `SliceIndex`. This is consistent with `get{,_mut}` and should make ranges work. It should also be more compatible with using smaller integer types to index into slices, if we ever get that to work.

```rust
fn get_many_mut<I, const N: usize>(
    &mut self,
    indices: [I; N],
) -> Result<[&mut <I as SliceIndex<[T]>>::Output; N], GetManyMutError<N>>
where
    // If we switch to using `DisjointIndices` as suggested above, `Ord` is not
    // needed here.
    I: SliceIndex<[T]> + Ord
{ /* ... */ }
```

Cc @ChayimFriedman2 since you have a stabilization PR up.

**Comment by tguichaoua** [comment]
> That would leave space open to also allow things like `get_many_mut((..4, 4, 5..))` in future, should we decide to allow it.

I create a [POC](https://github.com/tguichaoua/rust_slice_get_many_poc) to experiment with this idea.

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting yesterday and concluded that we are mostly happy with the shape that the API has today. We would like it to be extended to support any index that implements `SliceIndex`. We are deferring on tuple support for now.

**Comment by ChayimFriedman2** [comment]
@Amanieu Is the team interested in adding range support now, or deferring it?

**Comment by Amanieu** [comment]
Yes, we would like to add range support now.

**Comment by ChayimFriedman2** [comment]
@Amanieu Implementing this with `SliceIndex` will make adding support for tuples a breaking change, since users will be able to pass an opaque type that is `SliceIndex` into `get_many_mut()`.

**Comment by ChayimFriedman2** [comment]
@Amanieu Did you discuss whether the error should include more details?

**Comment by Amanieu** [comment]
> Did you discuss whether the error should include more details?

I'll nominate that for discussion in our next meeting.

**Comment by the8472** [comment]
> We'd like to stabilize this, using the signature that accepts an array of SliceIndex types.

Ah, I thought we were talking about a `many_mut(T) where T: ManyIdx`  here where `ManyIdx` would be implemented for arrays of ranges and arrays of usizes. Not an `[T;N] where T: ManyIdx` or such.

**Comment by ChayimFriedman2** [comment]
@the8472 Why do it? It both complicates the code and makes possible extension of supporting tuples a breaking change.

**Comment by the8472** [comment]
Eh? That's the opposite...

With `many_mut(T) where T: ManyIdx` you can support `impl ManyIdx for  (A, B)` and ``impl ManyIdx for [T; N]`. With `many_mut([T;N]) where T: ManyIdx` you can only ever pass arrays to many_mut, not tuples.

**Comment by ChayimFriedman2** [comment]
Oh nevermind it was only about `SliceIndex` but I use a different trait anyway. Still this is not blocking us from supporting tuples in the future by adding a new trait, and we likely want two traits anyway for better code reuse (although I didn't try implementing tuple support so I can't be sure). And libs-api said they want to defer tuple support anyway.

**Comment by the8472** [comment]
> Still this is not blocking us from supporting tuples in the future by adding a new trait

Maybe. At least it'd then need a crater run to change it from an array type to a trait since it could lead to inference failures.

**Comment by Amanieu** [comment]
I would actually argue against supporting tuples in the API: this *significantly* complicates the API and yet it doesn't enable doing anything in safe code that couldn't be done before. Any range type can be converted to a plain `Range` before it is used with a `Vec`, so you don't lose any functionality by requiring all range types to be the same.

**Comment by joshtriplett** [comment]
I share the opinion that we should not support tuples in `get_many_mut`. I think if we want to support tuples (in order to support different range types, such as `(..x, y..)`) we should add a separate `get_many_ranges_mut`, to avoid complicating the signature of `get_many_mut` any more than it is already.

**Comment by ChayimFriedman2** [comment]
> I would actually argue against supporting tuples in the API: this _significantly_ complicates the API and yet it doesn't enable doing anything in safe code that couldn't be done before. Any range type can be converted to a plain `Range` before it is used with a `Vec`, so you don't lose any functionality by requiring all range types to be the same.

Actually it loses a tiny (really tiny) bit of performance: checking whether two usizes overlap can be done with one comparison less than checking if two ranges overlap, so if you have e.g. two usizes and one range it will be one comparison less, and also, checking if a usize is out of bounds can be done with one comparison less than a range. LLVM may optimize `i..i+1` but it may fail because of possible overflow, and `i..=i` requires all ranges to be `RangeInclusive` which can require addition/subtraction.

**Comment by m-ou-se** [comment]
@rfcbot concern "many"

The main use case for this method is to get a small number (e.g. 2 or 3) items. For getting a _many_ items, this method can be quite inefficient, as it needs to check for duplicate indexes.

Putting "many" feels quite wrong for that reason, as it is meant for (and optimized for) 'few' items, not for 'many' items.

Should we name this `get_multiple_mut` or `get_n_mut` or `get_muts` or something else that doesn't imply it's purpose is to get 'many' items?

**Comment by the8472** [comment]
At least the unchecked variants should be efficient even for many items. We could also include optimized code paths for N > 16 or at whatever point optimizations cease to work.

**Comment by joshtriplett** [comment]
@m-ou-se I agree completely with the concern that `many` is misleading and doesn't steer people towards the primary use case of 2-3.

Every single use of `many` in the standard library is currently unstable, so I'd support renaming them all consistently.

:+1: for `get_n_mut`, which is really clear, short, and `n` is evocative.

:-1: for `get_multiple_mut` or any longer name.

**Comment by Xiretza** [comment]
I agree that `get_multiple_mut` is a little long (though I don't necessarily think too long), but I'm not a fan of `get_n_mut`. It feels to jargon-y to me.

**Comment by Amanieu** [comment]
I personally prefer the current `get_many_mut`: I don't think the confusion around "many" will be a problem in practice since you do have to statically provide the size of the array at compile-time. And I think that "many" clearly and concisely describes how this method differs from `get_mut`.

**Comment by crumblingstatue** [comment]
Another alternative is `get_multi_mut`, which is a little shorter than `get_multiple_mut`.

**Comment by deadsoul44** [comment]
What about:

`get_mut_many`
or
`get_mut_multi`

This way, developers can find both `get_mut` and `get_mut_many` or `get_mut_multi` when searching code.

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting today and decided to change the name to `get_multi_mut`, which was the least controversial choice. As such I will restart the FCP.

@rfcbot cancel

**Comment by Amanieu** [comment]
@rfcbot merge

**Comment by Amanieu** [comment]
> The error should be a two-variant enum (one variant for overlaps and one variant for out-of-bounds) that doesn't carry any data and doesn't have a generic parameter.

Also note that the error type still needs to be changed as we decided in a previous meeting.

**Comment by cuviper** [comment]
> We discussed this in the libs-api meeting today and decided to change the name to `get_multi_mut`, which was the least controversial choice.

Was something like `get_array_mut` considered? Since we are in fact returning an array...

**Comment by orlp** [comment]
In `slotmap` this functionality is called [`get_disjoint_mut`](https://docs.rs/slotmap/latest/slotmap/basic/struct.SlotMap.html#method.get_disjoint_mut).

**Comment by m-ou-se** [comment]
@rfcbot concern Consider `get_disjoint_mut` as the name.

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting and most of the team preferred `get_disjoint_mut` for both slices and `HashMap`. Restarting the FCP again!

@rfcbot cancel

**Comment by Amanieu** [comment]
@rfcbot merge
